Content.Server/GameObjects/Components/BaseComputerUserInterfaceComponent.cs

```diff
@@ -0,0 +1,144 @@
+        /// Calling base is necessary if you want this class to have any meaning.
+        /// </summary>
+        protected void OnReceiveUnfilteredUserInterfaceMessage(ServerBoundUserInterfaceMessage obj, IEntity sessionEntity)
+        {
+            // "Across all computers" "anti-cheats" ought to be put here or at some parent level (BaseDeviceUserInterfaceComponent?)
+            // Determine some facts about the session.
+            // Powered?
+            if (!Powered)
+            {
+                sessionEntity.PopupMessageCursor(Loc.GetString("base-computer-ui-component-not-powered"));
+                return; // Not powered, so this computer should probably do nothing.
+            }
+            // Can we interact?
+            if (!ActionBlockerSystem.CanInteract(sessionEntity))
+            {
+                sessionEntity.PopupMessageCursor(Loc.GetString("base-computer-ui-component-cannot-interact"));
+                return;
+            }
+            // Good to go!
+            OnReceiveUserInterfaceMessage(obj);
+        }
+
+        /// <summary>
+        /// Override this to handle messages from the UI.
+        /// Calling base is unnecessary.
+        /// These messages will automatically be blocked if the user shouldn't be able to access this computer, or if the computer has lost power.
+        /// </summary>
+        protected virtual void OnReceiveUserInterfaceMessage(ServerBoundUserInterfaceMessage obj)
+        {
+            // Nothing!
+        }
+
+        public override void HandleMessage(ComponentMessage message, IComponent? component)
+        {
+            base.HandleMessage(message, component);
+            switch (message)
+            {
+                case PowerChangedMessage powerChanged:
+                    PowerReceiverOnOnPowerStateChanged(powerChanged);
+                    break;
+            }
+        }
+
+        private void PowerReceiverOnOnPowerStateChanged(PowerChangedMessage e)
+        {
+            if (!e.Powered)
+            {
+                // We need to kick off users who are using it when it loses power.
+                UserInterface?.CloseAll();
+                // Now alert subclass.
+                ComputerLostPower();
+            }
+        }
+
+        /// <summary>
+        /// Override this if you want the computer to do something when it loses power (i.e. reset state)
+        /// All UIs should have been closed by the time this is called.
+        /// Calling base is unnecessary.
+        /// </summary>
+        public virtual void ComputerLostPower()
+        {
+        }
+
+        /// <summary>
+        /// This is called from ComputerUIActivatorSystem.
+        /// Override this to add additional activation conditions of some sort.
+        /// Calling base runs standard activation logic.
+        /// *This remains inside the component for overridability.*
+        /// </summary>
+        public virtual void ActivateThunk(ActivateInWorldEvent eventArgs)
```

Comment:
Doesn't this defeat the purpose of using the system if it's just calling component logic anyway?

Reply:
*This remains inside the component for overridability.*